### ISSUE TITLE
Bug Report: Different results from evalEngine and MySQL for dates on/before `0000-02-29`

### ISSUE THREAD

**Issue by beingnoble03** [vitessio/vitess#15077]
### Overview of the Issue

Currently, evalEngine considers `0000` as a leap year, while MySQL doesn't.
We are missing the test case of `0000-02-29` in the functions where Date is expected as an argument.

Also, when we pass dates before `0000-02-29` we obtain different results from MySQL and evalEngine. We need to fix this.

To fix this,
This function in `go/mysql/datetime/helpers.go` should be fixed.
```go
func isLeap(year int) bool {
	return year%4 == 0 && (year%100 != 0 || year%400 == 0)
}
```
and, inbuilt-mysql functions need to be fixed, because in some cases MySQL returns 0 instead of NULL when `0000-02-29`.

### Reproduction Steps

1. Add `"'0000-02-29'"` and `"'0000-01-03'"` these 2 cases in `inputConversions` present in go/vt/vtgate/evalengine/testcases/inputs.go
2. Run `go test ./go/vt/vtgate/evalengine/integration -run TestMySQL -v`
3. View the errors

### Binary Version

```sh
Version: 19.0.0
```

### Operating System and Environment details

```sh
- Operating system macOS 14.0
- Kernel version Darwin 23.0.0
- Architecture arm64
```

**Comment by beingnoble03** [comment]
Please let me know if we need to fix this.

**Comment by vmg** [comment]
Yeah this report looks legit. It seems like you already have a good idea of what needs to be fixed. Could you please open a PR? Thank you!

**Comment by beingnoble03** [comment]
Yeah sure, I'll open a PR.
